wrangler-service/src/main/java/io/cdap/wrangler/service/directive/WorkspaceHandler.java

```diff
@@ -415,4 +409,79 @@ private String generateWorkspaceName(WorkspaceId id, @Nullable String path) {
+    directives.removeIf(d -> PRAGMA_PATTERN.matcher(d).find());
+
+    return getContext().isRemoteTaskEnabled() ?
+      executeRemotely(namespace, directives, detail) :
+      executeLocally(namespace, directives, detail);
+  }
+
+  /**
+   * Executes the given list of directives on the given workspace locally in the same JVM.
+   *
+   * @param namespace the namespace to operate on for finding user defined directives
+   * @param directives the list of directives to apply. The list provided must be a mutable list for the addition of
+   *                   {@code #pragma} directives for loading UDDs.
+   * @param detail the workspace to operate on
+   * @return the resulting rows after applying the directives
+   */
+  private List<Row> executeLocally(String namespace, List<String> directives,
+                                   WorkspaceDetail detail) throws DirectiveLoadException, DirectiveParseException {
+    // load the udd
+    composite.reload(namespace);
+    return executeDirectives(namespace, directives, new ArrayList<>(detail.getSample()));
+  }
+
+  /**
+   * Executes the given list of directives on the given workspace remotely using the task worker framework.
+   *
+   * @param namespace the namespace to operate on for finding user defined directives
+   * @param directives the list of directives to apply. The list provided must be a mutable list for the addition of
+   *                   {@code #pragma} directives for loading UDDs.
+   * @param detail the workspace to operate on
+   * @return the resulting rows after applying the directives
+   */
+  private List<Row> executeRemotely(String namespace, List<String> directives,
+                                    WorkspaceDetail detail) throws Exception {
+
+    GrammarMigrator migrator = new MigrateToV2(directives);
+    String recipe = migrator.migrate();
+    Map<String, DirectiveClass> systemDirectives = new HashMap<>();
+
+    // Gather system directives and add #pragma for UDD.
+    Set<String> userDirectives = new HashSet<>();
+    GrammarWalker walker = new GrammarWalker(new RecipeCompiler(), new ConfigDirectiveContext(DirectiveConfig.EMPTY));
+    walker.walk(recipe, (command, tokenGroup) -> {
+      DirectiveInfo info = SystemDirectiveRegistry.INSTANCE.get(namespace, command);
+      if (info != null) {
+        systemDirectives.put(command, info.getDirectiveClass());
+      } else {
+        userDirectives.add(command);
+      }
+    });
+    if (!userDirectives.isEmpty()) {
+      directives.add(0, "#pragma load-directives " + String.join(",", userDirectives) + ";");
+    }
+
+    RemoteDirectiveRequest directiveRequest = new RemoteDirectiveRequest(recipe, systemDirectives,
+                                                                         namespace, detail.getSampleAsBytes());
+    RunnableTaskRequest runnableTaskRequest = RunnableTaskRequest.getBuilder(RemoteExecutionTask.class.getName())
```

Comment:
Will it help to send the list of commands in RemoteDirectiveRequest? Since we are already compiling and walking the recipe, looks like we have all the information already required for the directives here. This will eliminate the need for using GrammarWalker again in remote execution.

Reply:
I tried to carry the parsed result over to the remote task, but turns out it is quite complex to serialize the `Token` interface, which has implementations for various type. The tree walking time is actually fast comparing to other operations (10s of milliseconds only).

Reply:
Sounds good.